File: packages/zod/src/notification.ts
import { z } from "./z.js";

// --- Notification Schemas ---

export const NotificationSchema = z.object({
    id: z.string().uuid(),
    user_id: z.string().uuid(),
    actor_id: z.string().uuid().nullable(),
    project_id: z.string().uuid().nullable(),
    issue_id: z.string().uuid().nullable(),
    type: z.string(),
    title: z.string(),
    message: z.string(),
    payload: z.any(), // json.RawMessage in Go
    is_read: z.boolean(),
    created_at: z.string(),
}).openapi("Notification");

export const ListNotificationsRequestSchema = z.object({
    is_read: z.boolean().optional(),
    limit: z.coerce.number().min(1).max(100).optional().default(20),
    offset: z.coerce.number().min(0).optional().default(0),
});

export const MarkAsReadSchema = z.object({
    id: z.string().uuid(),
});

// --- Type Exports ---

export type Notification = z.infer<typeof NotificationSchema>;
export type ListNotificationsRequest = z.infer<typeof ListNotificationsRequestSchema>;
export type MarkAsRead = z.infer<typeof MarkAsReadSchema>;

File: backend/internal/model/notification.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// Notification represents a notification sent to a user.
type Notification struct {
	ID        uuid.UUID       `json:"id" db:"id"`
	UserID    uuid.UUID       `json:"user_id" db:"user_id"`
	ActorID   *uuid.UUID      `json:"actor_id" db:"actor_id"`
	ProjectID *uuid.UUID      `json:"project_id" db:"project_id"`
	IssueID   *uuid.UUID      `json:"issue_id" db:"issue_id"`
	Type      string          `json:"type" db:"type"`
	Title     string          `json:"title" db:"title"`
	Message   string          `json:"message" db:"message"`
	Payload   json.RawMessage `json:"payload" db:"payload"`
	IsRead    bool            `json:"is_read" db:"is_read"`
	CreatedAt time.Time       `json:"created_at" db:"created_at"`
}

// CreateNotificationParams defines the input for creating a new notification.
type CreateNotificationParams struct {
	UserID    uuid.UUID       `json:"user_id"`
	ActorID   *uuid.UUID      `json:"actor_id"`
	ProjectID *uuid.UUID      `json:"project_id"`
	IssueID   *uuid.UUID      `json:"issue_id"`
	Type      string          `json:"type"`
	Title     string          `json:"title"`
	Message   string          `json:"message"`
	Payload   json.RawMessage `json:"payload"`
}

// NotificationFilters defines filtering options for listing notifications.
type NotificationFilters struct {
	UserID *uuid.UUID `json:"user_id"`
	IsRead *bool      `json:"is_read"`
	Limit  int        `json:"limit"`
	Offset int        `json:"offset"`
}
